src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,139 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int elementsNumber;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> current = new Node<T>(tail, value, null);
+        if (elementsNumber == 0) {
+            head = current;
+        } else {
+            tail.next = current;
+        }
+        tail = current;
+        elementsNumber++;
+        return true;
+
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == elementsNumber) {
+            add(value);
+            return;
+        }
+        checkElementIndex(index);
+        Node<T> oldNode = node(index);
+        Node<T> newNode = new Node<>(oldNode.previous, value, oldNode);
+        if (index > 0) {
+            newNode.previous.next = newNode;
+        } else {
+            head = newNode;
+        }
+        oldNode.previous = newNode;
+        elementsNumber++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> x = node(index);
+        T oldVal = x.element;
+        x.element = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = node(index);
+        if (index > 0) {
+            node.previous.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.previous = node.previous;
+        } else {
+            tail = node.previous;
+        }
+        T removedItem = node.element;
+        elementsNumber--;
```

Comment:
let's move this code to separate method `unlink(Node<T> nodeToRemove)` and call it here and in the method remove(T t)

Reply:
wasn't able to add it to remove method, seems that it returns the same as remove, but haven't worked out